clang/lib/CIR/Dialect/Transforms/GotoSolver.cpp

```diff
@@ -0,0 +1,57 @@
+#include "PassDetail.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "clang/CIR/Dialect/IR/CIRDialect.h"
+#include "clang/CIR/Dialect/Passes.h"
+
+using namespace mlir;
+using namespace mlir::cir;
+
+namespace {
+
+struct GotoSolverPass : public GotoSolverBase<GotoSolverPass> {
+
+  GotoSolverPass() = default;
+  void runOnOperation() override;
+};
+
+static void process(mlir::cir::FuncOp func) {
+
+  mlir::OpBuilder rewriter(func.getContext());
+  std::map<std::string, Block *> labels;
+  std::vector<mlir::cir::GotoOp> gotos;
+
+  func.getBody().walk([&](mlir::Operation *op) {
+    if (auto lab = dyn_cast<mlir::cir::LabelOp>(op)) {
+      labels.emplace(lab.getLabel().str(), lab->getBlock());
+      lab.erase();
+    } else if (auto goTo = dyn_cast<mlir::cir::GotoOp>(op)) {
+      gotos.push_back(goTo);
+    }
+  });
+
+  for (auto goTo : gotos) {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(goTo);
+    auto dest = labels[goTo.getLabel().str()];
+    rewriter.create<mlir::cir::BrOp>(goTo.getLoc(), dest);
+    goTo.erase();
+  }
+}
+
+void GotoSolverPass::runOnOperation() {
+  SmallVector<Operation *, 16> ops;
+  getOperation()->walk([&](mlir::cir::FuncOp op) {
+    if (op.getHasLabels())
```

Comment:
By removing the attribute, you could do a `FuncOp->walk([&](mlir::cir::LabelOp op)` here, and as part of that already collect all labels.

From the tracked labels, you could get symbol uses, which will give you all the gotos (since they should be verified to always point to an existing label).

If you do it like this, you only walk the function once.